compiler/damlc/daml-preprocessor/src/DA/Daml/Preprocessor/TemplateConstraint.hs

```diff
@@ -0,0 +1,83 @@
+-- Copyright (c) 2019 The DAML Authors. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+-- | Every `template T a_1 a_n` gets desugared into a `class TInstance a_1 ... a_n`,
+-- an instance
+-- ```
+-- (*)    instance TInstance a_1 ... a_n => Template (T a_1 ... a_n)
+-- ```
+-- and instances `instance TInstance a_1 ... a_n => Choice (T a_1 ... a_n) C R`
+-- for each choice `C`.
+--
+-- Thus, a _generic_ exercise of a choice on `T t_1 ... t_n` needs to have the
+-- `TInstance t_1 ... t_n` constraint in scope. However, we want to keep the
+-- existence of the `TInstance` class an implementation detail and not expose
+-- it to our users. Instead we want our users to add a `Template (T t_1 ... t_n)`
+-- constraint when they want to perform the generic exercise.
+--
+-- Due to the (*) instance above, the constraint `Template (T t_1 ... t_n)`
+-- is satisfied if and only if the constraint `TInstance t_1 ... t_n` is
+-- satisfied. For the intent described above it would be necessary that GHC
+-- conlcudes the latter from the former. Unfortunately, GHC's type system only
+-- allows for concluding the former from the latter.
+--
+-- Thus, we add a preprocessing step which rewrites all constraints of
+-- the form `Template (T t_1 ... t_n)` into `TInstance t_1 ... t_n`.
+module DA.Daml.Preprocessor.TemplateConstraint (
+    templateConstraintPreprocessor
+    ) where
+
+import "ghc-lib" GHC
+import "ghc-lib-parser" BasicTypes
+import "ghc-lib-parser" OccName
+import "ghc-lib-parser" RdrName
+
+import Data.Generics.Uniplate.Data
+
+templateConstraintPreprocessor :: ParsedSource -> ParsedSource
+templateConstraintPreprocessor = fmap onModule
+
+onModule :: HsModule GhcPs -> HsModule GhcPs
+onModule = transformBi onType . transformBi onTyClDecl
+
+-- | The contexts of class and data definitions do not live inside an
+-- `HsType` but in their respective AST nodes. That's why we need to chase
+-- them down separately.
```

Comment:
`HsContext` is a synonym for `[HsType]` as I remember - does Uniplate `descend` not understand that and descend on the contexts?

Reply:
Uniplate works type based. Thus, it would descend into all `[HsType]` occurrences. I don't know if they appear somewhere else now or might do so in the future. That's why I prefer to be precise about this.